### PR TITLE
feat(aide): Added `NoApi` and `WithApi` utility structs

### PR DESCRIPTION
**Pull request by Wicpar**
WIP

Added first draft of `NoAPI` and `WithApi` Utility structs as discussed in #63 
Todo:
- [x] Documentation
- [ ] Tests

### CODE REVIEW

**Comment by netthier** [comment]
I'd really appreciate this feature going through, since it effectively makes aide infinitely extensible without having to implement the operation IO traits manually. Is there anything I can do to help? :thinking:

**Comment by Wicpar** [comment]
It just needs a rewrite of existing implementations with this system.
I still wanted to see if anyone had input on how to improve it though, but we can push it through like this.

**Comment by Wicpar** [comment]
I think we should limit ourselves to implementing types that need `WithApi` and not NoApi, and just make an example for NoApi uses `fn endpoint (NoApi(mut tx): NoApi<Tx<Postgres>>)`

File: crates/aide/src/helpers/no_api.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    Hash,
+    Serialize,
+    Deserialize,
+    Deref,
+    DerefMut,
+    AsRef,
+    AsMut,
+    From,
+)]
+pub struct NoApi<T>(pub T);
```
Comment (tamasfe) [inline crates/aide/src/helpers/no_api.rs:31]: Couldn't this be simplified with something like below?
```rs
pub type NoApi<T> = WithApi<T, SomeTypeThatHasNoDocs>;
```

Comment (Wicpar) [reply crates/aide/src/helpers/no_api.rs:31]: Yes indeed

Comment (Wicpar) [reply crates/aide/src/helpers/no_api.rs:31]: Actually not, unformtunately rust requires the phantom data so
```
fn endpoint (NoApi(mut tx): NoApi<Tx<Postgres>>)
```
becomes
```
fn endpoint (NoApi(mut tx, _): NoApi<Tx<Postgres>>)
```

**Comment by tamasfe** [comment]
> I still wanted to see if anyone had input on how to improve it though, but we can push it through like this.

It looks good to me, thanks for the effort!

> It just needs a rewrite of existing implementations with this system.

This shouldn't be a blocker, should it? Technically API-wise they could be breaking changes, but not something end-users would notice (I assume).

> Maybe even split the features into separate crates so that multiple versions can be selected for crates like axum-sqlx-tx and then reexport the relevant version with features.

After this is merged I'm inclined to drop support for less popular libraries that are not part of the main frameworks altogether (e.g. everything outside `axum` or `axum-*` and common stuff like `bytes`), as they're potentially infinite burden for us.

`aide` should be a slowly moving target, so if someone writes a library, they can choose to support `aide`, not the other way around. If a library doesn't support `aide`, with this PR the end user should be able to work around that more or less seamlessly as well.

**Comment by Wicpar** [comment]
Yes the merge without the changes should be ok.

The only issue i still have on my side with this crate is `axum-sqlx-tx` having the wrong version, if we can remove that and other empty `OperationInput`/`OperationOutput` in favor of the `NoApi` syntax at some point that would be good.

At some point the way to do things would be to reexport NoApi<Something> as Something an dthat can be packaged i side crates even if the main library doesn't implement aide.

File: crates/aide/src/helpers/no_api.rs (hunk 1 above)
Comment (tamasfe) [reply crates/aide/src/helpers/no_api.rs:31]: Welp, yeah I see no way around that either, we could suggest using `.into_inner()` in downstream code, but the appearing extra field out of nowhere can be confusing indeed.

Comment (Wicpar) [reply crates/aide/src/helpers/no_api.rs:31]: The proper Dry code would be using a macro, but it's really not worth the horrible proc macro syntax for two impls, it's not like there is a lot of code duplication.

Comment (Wicpar) [reply crates/aide/src/helpers/no_api.rs:31]: maybe we could do something with the axum impls, they are quite similar... Maybe utility trait + macro ? but it's really complexifying things

**Comment by Wicpar** [comment]
I successfully migrated my axum-sqlx-tx code as follows mith the master branch:

```rust
use aide::NoApi;
use axum_sqlx_tx::Tx as BaseTx;

pub type Tx<T> = NoApi<BaseTx<T>>;
```

```rust
#[debug_handler(state = ServerState)]
async fn get_api_keys(
    mut tx: Tx<Postgres>, // previously defined tx
    session: Session,
) -> Result<Json<Vec<ApiKey>>, StatusError> {
    Ok(Json(
        query!("blablabla")
        .fetch_all(&mut *tx) // deref mut
        .await?,
    ))
}
```

**Pull request merged**